pkg/kubelet/server.go

```diff
@@ -620,80 +628,179 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+			connectionID = strconv.Itoa(int(stream.Identifier()))
+		case api.StreamTypeData:
+			connectionID = strconv.Itoa(int(stream.Identifier()) - 2)
+		default:
+			connectionID = ""
+		}
+
+		glog.V(5).Infof("Automatically assigning connection ID=%s from stream type=%s, ID=%d", connectionID, streamType, stream.Identifier())
+	}
+	return connectionID
+}
+
+// ServePortForward handles a port forwarding request.
+func ServePortForward(w http.ResponseWriter, req *http.Request, portForwarder PortForwarder, podName string, uid types.UID, idleTimeout time.Duration) {
+	streamChan := make(chan httpstream.Stream, 1)
+
+	glog.V(5).Infof("Upgrading port forward response")
+	upgrader := spdy.NewResponseUpgrader()
+	conn := upgrader.UpgradeResponse(w, req, portForwardStreamReceived(streamChan))
 	if conn == nil {
 		return
 	}
 	defer conn.Close()
-	conn.SetIdleTimeout(s.host.StreamingConnectionIdleTimeout())
 
-	var dataStreamLock sync.Mutex
-	dataStreamChans := make(map[string]chan httpstream.Stream)
+	glog.V(5).Infof("Setting port forwarding streaming connection idle timeout to %v", idleTimeout)
+	conn.SetIdleTimeout(idleTimeout)
 
+	var streamLock sync.Mutex
+	streamChans := make(map[string]chan httpstream.Stream)
+
+	glog.V(5).Infof("Waiting for port forward streams")
 Loop:
 	for {
 		select {
 		case <-conn.CloseChan():
+			glog.V(5).Infof("Upgraded connection closed")
 			break Loop
 		case stream := <-streamChan:
-			streamType := stream.Headers().Get(api.StreamType)
-			port := stream.Headers().Get(api.PortHeader)
-			dataStreamLock.Lock()
-			switch streamType {
-			case "error":
-				ch := make(chan httpstream.Stream)
-				dataStreamChans[port] = ch
-				go waitForPortForwardDataStreamAndRun(kubecontainer.GetPodFullName(pod), uid, stream, ch, s.host)
-			case "data":
-				ch, ok := dataStreamChans[port]
-				if ok {
-					ch <- stream
-					delete(dataStreamChans, port)
-				} else {
-					glog.Errorf("Unable to locate data stream channel for port %s", port)
-				}
-			default:
-				glog.Errorf("streamType header must be 'error' or 'data', got: '%s'", streamType)
-				stream.Reset()
+			connectionID := getPortForwardStreamConnectionID(stream)
+			glog.V(5).Infof("(%s) Received new stream", connectionID)
+
+			streamLock.Lock()
+
+			ch, ok := streamChans[connectionID]
+			if !ok {
+				glog.V(5).Infof("(%s) connection ID not in streamChans", connectionID)
+
+				glog.V(5).Infof("(%s) creating stream channel", connectionID)
+				ch = make(chan httpstream.Stream, 2)
+				streamChans[connectionID] = ch
+
+				glog.V(5).Infof("(%s) starting handlePortForwardStreams", connectionID)
+				go func() {
+					handlePortForwardStreams(podName, uid, ch, portForwarder, connectionID)
+					streamLock.Lock()
+					defer streamLock.Unlock()
+					delete(streamChans, connectionID)
+					close(ch)
+				}()
 			}
-			dataStreamLock.Unlock()
+
+			streamLock.Unlock()
+
+			glog.V(5).Infof("(%s) sending stream to stream channel", connectionID)
+			ch <- stream
```

Comment:
I'm still reading through, but it's not immediately obvious to me that this is guaranteed to happen before the close on this channel on line 734. If this writes to a closed channel, it will panic.